crates/sui-source-validation-service/tests/tests.rs

```diff
@@ -63,6 +144,103 @@ Multiple source verification errors found:
+    let resp = SuiClientCommands::Publish {
+        package_path: package_path.clone(),
+        build_config,
+        gas: Some(gas_obj_id),
+        gas_budget: rgp * TEST_ONLY_GAS_UNIT_FOR_PUBLISH,
+        skip_dependency_verification: false,
+        with_unpublished_dependencies: false,
+        serialize_unsigned_transaction: false,
+        serialize_signed_transaction: false,
+        lint: false,
+    }
+    .execute(context)
+    .await?;
+
+    let SuiClientCommandResult::Publish(response) = resp else {
+        unreachable!("Invalid response");
+    };
+    let SuiTransactionBlockEffects::V1(effects) = response.effects.unwrap();
+    assert!(effects.status.is_ok());
+    Ok(effects)
+}
+
+async fn run_upgrade(
+    upgrade_pkg_path: PathBuf,
+    cap: &OwnedObjectRef,
+    context: &mut WalletContext,
+    gas_obj_id: ObjectID,
+    rgp: u64,
+) -> anyhow::Result<()> {
+    let build_config = BuildConfig::new_for_testing().config;
+    let resp = SuiClientCommands::Upgrade {
+        package_path: upgrade_pkg_path,
+        upgrade_capability: cap.reference.object_id,
+        build_config,
+        gas: Some(gas_obj_id),
+        gas_budget: rgp * TEST_ONLY_GAS_UNIT_FOR_PUBLISH,
+        skip_dependency_verification: false,
+        with_unpublished_dependencies: false,
+        serialize_unsigned_transaction: false,
+        serialize_signed_transaction: false,
+        lint: false,
+    }
+    .execute(context)
+    .await?;
+
+    let SuiClientCommandResult::Upgrade(response) = resp else {
+            unreachable!("Invalid upgrade response");
+        };
+    let SuiTransactionBlockEffects::V1(effects) = response.effects.unwrap();
+    assert!(effects.status.is_ok());
+    Ok(())
+}
+
+/// Copy the package and set `published-at` in the Move toml file. The need for
+/// this will be subsumed by automated address management.
+fn copy_with_published_at_manifest(
+    source_path: &PathBuf,
+    dest_path: &PathBuf,
+    package_id: ObjectID,
+) -> PathBuf {
+    fs_extra::dir::copy(
+        source_path,
+        dest_path,
+        &fs_extra::dir::CopyOptions::default(),
+    )
+    .unwrap();
+    let mut upgrade_pkg_path = dest_path.clone();
+    upgrade_pkg_path.extend(["custom", "Move.toml"]);
+    let mut move_toml = std::fs::File::options()
+        .read(true)
+        .write(true)
+        .open(&upgrade_pkg_path)
+        .unwrap();
+    upgrade_pkg_path.pop();
+
+    let mut buf = String::new();
+    move_toml.read_to_string(&mut buf).unwrap();
+
+    // Add a `published-at = "0x<package_object_id>"` to the Move manifest.
+    let mut lines: Vec<String> = buf.split('\n').map(|x| x.to_string()).collect();
+    let idx = lines.iter().position(|s| s == "[package]").unwrap();
+    lines.insert(
+        idx + 1,
+        format!("published-at = \"{}\"", package_id.to_hex_uncompressed()),
+    );
+    let new = lines.join("\n");
+    move_toml.write_at(new.as_bytes(), 0).unwrap();
```

Comment:
Might be slightly neater to write the lines straight into the file:

```suggestion
    move_toml.set_len(0);
    for line in buf.lines() {
        write!(move_toml, "{line}\n").unwrap();
        if line == "[package]" {
            write!(move_toml, r#"published-at = "{}""#, package_id.to_hex_uncompressed()).unwrap();
        }
    }
```

Reply:
Thanks for the suggestion. I tried substituting the change mechanically and ran into a test error

```
    unexpected character found: `\u{0}` at line 1 column 1.
```

So I'll look into it on a follow up